### PR TITLE
Feat/329/jsonld validation

### PR DESCRIPTION
**Pull request by chunningham**
key files are `ts/linkedData/index.ts` and `ts/linkedData/types.ts`. Other changes are including jsonLD types in DID doc sections and credential types, and using the digestion/validation functions in DID doc and signed credential classes.

### CODE REVIEW

File: ts/linkedData/index.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ *  the '@context' on the data will be discarded.
+ * @param data - {@link JsonLdObject} without the '@context' section
+ * @param context - JsonLD context to use during normalization
+ */
+
+export const normalizeJsonLd = async (
+  { ['@context']: _, ...data }: JsonLdObject,
+  context: JsonLdContext,
+) => canonize(data, {
+  expandContext: context,
```
Comment (VolkerSchiewe) [inline ts/linkedData/index.ts]: replacing the context can always result in breaking the signature validation. This is why I'm afraid of this context expansion.

File: ts/linkedData/types.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import { ILinkedDataSignatureAttrs } from '../linkedDataSignature/types'
+
+type JsonLdPrimitive = string | number | boolean | JsonLdObject | JsonLdObject[]
+
+export type JsonLdContextEntry = string | JsonLdObject
+
+export type JsonLdContext = JsonLdContextEntry | JsonLdContextEntry[]
+
+export interface JsonLdObject {
+  '@context'?: JsonLdContext
```
Comment (VolkerSchiewe) [inline ts/linkedData/types.ts]: The context is the key element of a JsonLD, isn't it? Hence I do not see why this is an optional parameter

Comment (chunningham) [reply ts/linkedData/types.ts]: contexts are the killer app, but they're not actually mandated by the spec. any valid JSON object is also a valid JSON-LD object, context or not

File: ts/linkedData/index.ts (hunk 1 above)
Comment (chunningham) [reply ts/linkedData/index.ts]: context is provided as a second param so that contexts can be used for objects which may not have their own, e.g. normalizing a json-ld proof. An overload with just one arg which doesn't do this replacement may be good though

Comment (VolkerSchiewe) [reply ts/linkedData/index.ts]: Yes, replacing the context might be a nice feature for now (because may contexts are just garbage), but for long term we should avoid this... I would vote for a second optional parameter to override the context and use the existing one otherwise.

File: ts/linkedData/index.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      json.proof.creator,
+      issuerIdentity.didDocument,
+    )
+    return SoftwareKeyProvider.verify(
+      await digestJsonLd(json, json['@context']),
+      issuerPublicKey,
+      Buffer.from(json.proof.signatureValue, 'hex'),
+    )
+  } catch {
+    return false
```
Comment (Exulansis) [inline ts/linkedData/index.ts]: False can mean either failed validation, or failed resolution (i.e. no internet). Should this function throw in case resolution fails?

*Note: validateDigestable attempts to resolve the signer outside of a try / catch block.*

File: ts/linkedData/index.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+/**
+ * Helper function to handle JsonLD normalization.
+ * @dev The function expects the JsonLD '@context' to be passed as an argument,
+ *  the '@context' on the data will be discarded.
+ * @param data - {@link JsonLdObject} without the '@context' section
+ * @param context - JsonLD context to use during normalization
+ */
+
+export const normalizeJsonLd = async (
```
Comment (Exulansis) [inline ts/linkedData/index.ts]: Mark as @internal? Also applies to other functions exported for internal usage but not meant to be used directly.

File: ts/linkedData/index.ts (hunk 3 above)
Comment (Exulansis) [reply ts/linkedData/index.ts]: Also does this need unit tests?

File: ts/utils/contexts.ts [hunk 5]
```diff
@@ -1,8 +1,8 @@
-import { ContextEntry } from 'cred-types-jolocom-core'
+import { JsonLdContextEntry } from '../linkedData/types'
 
 /* Expanded default context for verifiable credentials, more verbose, but works in offline use cases */
 
-export const defaultContext: ContextEntry[] = [
+export const defaultContext: JsonLdContextEntry[] = [
```
Comment (Exulansis) [inline ts/utils/contexts.ts]: The type can be inferred by TS here, which allows us to not export the `JsonLdContextEntry` type at all, and only use the more general `JsonLdContext` type.

File: tests/data/didDocument.data.ts [hunk 6]
```diff
@@ -40,6 +40,7 @@ export const didDocumentJSON = {
 export const normalizedDidDocument =
```
Comment (Exulansis) [inline tests/data/didDocument.data.ts]: Some things are missing in the normalized output, i.e. the types of the keys in the `publicKey` and `authentication` sections.

File: tests/data/didDocument.data.ts [hunk 7]
```diff
@@ -40,6 +40,7 @@ export const didDocumentJSON = {
...
   '\
 <did:jolo:b2d5d8d6cc140033419b54a237a5db51710439f9f462d1fc98f698eca7ce9777#keys-1> <https://w3id.org/security#owner> <did:jolo:b2d5d8d6cc140033419b54a237a5db51710439f9f462d1fc98f698eca7ce9777> .\n\
+<did:jolo:b2d5d8d6cc140033419b54a237a5db51710439f9f462d1fc98f698eca7ce9777#keys-1> <https://w3id.org/security#publicKeyHex> "03848af62bffceb57631780ac0e0726106ee1c23262d6fd7ef906559d68f53a551" .\n\
```
Comment (VolkerSchiewe) [inline tests/data/didDocument.data.ts]: Why is this line added? you didn't changed the context, did you?
If so, this is definitely a breaking change

Comment (Exulansis) [reply tests/data/didDocument.data.ts]: Previously this bit wasn't tested correctly. If you look at how `normalizedDidDocument` is used on develop, you'll see it was only as output for the stubbed `canonize` mock, so we were expecting a mocked function to return the output it was mocked with more or less.

Comment (chunningham) [reply tests/data/didDocument.data.ts]: the stubbing of the normalization was removed from the did doc tests, so the test actually expands the context now. the publicKeyHex line should have appeared in earlier tests as it is present in both the context and the didDoc
